pandas-stubs/core/groupby/generic.pyi

```diff
@@ -149,6 +150,9 @@ class SeriesGroupBy(GroupBy, Generic[S1]):
 class _DataFrameGroupByScalar(DataFrameGroupBy):
     def __iter__(self) -> Iterator[tuple[Scalar, DataFrame]]: ...
 
+class _DataFrameGroupByPeriod(DataFrameGroupBy):
```

Comment:
If the only difference between these many helper classes is `__iter__`, it might be nicer to have one Generic class:

```py
class _DataFrameGroupBy(DataFrameGroupBy, Generic[T]):
    def __iter__(self) -> Iterator[tuple[T, DataFrame]]: ...
```

@Dr-Irv

Reply:
I think that's a good idea.  Would need to create a type alias (ideally prefixed with an underscore and only in this file - not sure if it would require to be declared in `_typing.py`) that is a `TypeVar` that includes `Scalar`, `Period`, and `tuple` and then modify `DataFrame.apply()` to return the correct generic version.

Reply:
I also think a generic is a good idea, although there is a slightly higher potential that it will break people's code, it's a lot better than pretending there are subclasses of `GroupByDataFrame`, which comes with its own semantic problems.

Putting the `TypeVar` into `_typing,py` seems sensible, if anyone creates a function that operates on a `GroupByDataFrame`  and returns a modified version, you probably want them to be able to adjust the signature so the bound type is preserved, so they would need to either redefine the `TypeVar` or import/access it in a type checking context from `pandas.core.groupby.generic` which is a lot more verbose than `pandas._typing`.